documentation/Advanced Features/Objects and Extensions.md

```diff
@@ -0,0 +1,269 @@
+"memberName" <value> ŀ
+```
+
+This sets the value of the member to the value on the stack. If the member is a
+restriced or private member, this will fail, unless inside an extension method.
+
+## Example
+
+```
+#:O Map | 
+  #[#] #!keys
+  #[#] #!values
+}
+
+λ value, key, mp | 
+  #$mp "keys" ᵇ« #$key ŀ 
+  #$mp "values" ᵇ« #$value ŀ
+  #$mp
+} #=put
+
+#$Map Ė
+"key1" "value1" #$put Ė
+```
+
+This example creates an object that acts like a map. It has two public members,
+`keys` and `values`, which are lists of the keys and values in the map. It then
+creates a lambda that takes a value, a key, and a map, and adds the value and key
+to the map. Finally, it creates a new map, and adds a key and value to it.
+
+Note that the lambda is not a member of the object, nor is it linked to the
+object in any way. It is simply a lambda that takes a map as an argument.
+
+## Extension Methods
+
+In the previous example, a lambda was used to add a key and value to a `Map` object.
+However, the lambda is in no way linked to the `Map` object - it can be called
+with any object, and will attempt to retrieve the `keys` and `values` members.
+This obviously is not desirable behaviour, even for a language as loose with types
+as Vyxal. To solve this, extension methods can be used to create fine-grained
+control over the types involved in a function.
+
+To define an extension method:
+
+```
+#:>> symbol | argA | typeA | argB | typeB | ... | argN | typeN | implementation }
+```
+
+* `symbol` is the symbol that will be used to call the extension method. This can
+be any valid symbol allowed in a custom element.
+* `arg` is the name of an argument to the extension method.
+* `type` is the type of an argument to the extension method. This can be any valid type name, as well as `*`, which indicates an any type.
+* `implementation` is the implementation of the extension method. This will always
+be the last branch.
+
+To use an extension method, call it as you would any custom element:
+
+```
+#:@symbol
+```
+
+### Important Notes
+
+* Extension methods require at least one argument. Attempting to define an
+extension method with no arguments will result in an error. Use the define
+structure instead.
+* Extension methods will be scoped globally. Currently, there is no way to define
+an extension method that is only available within a certain scope/context.
+* The types of the arguments are checked at runtime. If the types do not match,
+an error will be thrown.
+* Like custom elements, the implementation branch is where the actual code for the element or modifier is written. When called, the method will pop all arguments from the stack
+and execute the implementation as if it were a lambda. The implementation branch can be any valid Vyxal code.
+
+### Order of Extension Method Matching
+
+When a symbol is called, first the interpreter will check for an extension method
+with the appropriate types of items on the stack. If found, it will execute the

```

Comment:
Might be worth mentioning the phrases "multimethods" or "multiple dispatch" somewhere, to help people who already understand those concepts (as far as I can tell, "multiple dispatch" is an accurate way of describing the type-based overloading Vyxal does). Maybe also throw in a bit about *why* we do type-based overloading (better to have the language do it for you than do it manually, safer than languages like Python which let you do whatever, blah blah blah). Maybe mention Julia?